src/D2L.CodeStyle.Analyzers/CustomerState/CustomerStateAnalyzer.cs

```diff
@@ -0,0 +1,164 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using D2L.CodeStyle.Analyzers.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.CustomerState {
+
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	public sealed class CustomerStateAnalyzer : DiagnosticAnalyzer {
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+			Diagnostics.SingletonDependencyHasCustomerState
+		);
+
+		public override void Initialize(
+			AnalysisContext context
+		) {
+			//context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction( RegisterAnalysis );
+		}
+
+		private void RegisterAnalysis(
+			CompilationStartAnalysisContext context
+		) {
+			var singletonAttribute = context.Compilation.GetTypeByMetadataName(
+				"D2L.LP.Extensibility.Activation.Domain.SingletonAttribute" );
+
+			var customerStateAttribute = context.Compilation.GetTypeByMetadataName(
+				"D2L.LP.Extensibility.Activation.Domain.CustomerStateAttribute" );
+
+			if( singletonAttribute == null
+				|| singletonAttribute.Kind == SymbolKind.ErrorType
+			) {
+				return;
+			}
+
+			if( customerStateAttribute == null
+				|| customerStateAttribute.Kind == SymbolKind.ErrorType
+			) {
+				return;
+			}
+
+			context.RegisterSyntaxNodeAction(
+				ctx => AnalyzeClass(
+					ctx,
+					singletonAttribute,
+					customerStateAttribute
+				),
+				SyntaxKind.ClassDeclaration
+			);
+		}
+
+		private void AnalyzeClass(
+			SyntaxNodeAnalysisContext context,
+			INamedTypeSymbol singletonAttribute,
+			INamedTypeSymbol customerStateAttribute
+		) {
+			var root = context.Node as ClassDeclarationSyntax;
+			if( root == null ) {
+				return;
+			}
+
+			var symbol = context.SemanticModel.GetDeclaredSymbol( root );
+			if( symbol == null ) {
+				return;
+			}
+
+			// skip classes not marked singleton
+			if( !symbol.IsTypeMarkedSingleton() ) {
+				return;
+			}
+
+			var inspectedTypes = new HashSet<ITypeSymbol>();
+			AnalyzeType( context, root, symbol, customerStateAttribute, inspectedTypes );
+		}
+
+		private void AnalyzeType(
+			SyntaxNodeAnalysisContext context,
+			ClassDeclarationSyntax root,
+			ITypeSymbol type,
+			INamedTypeSymbol customerStateAttribute,
+			HashSet<ITypeSymbol> inspectedTypes
+		) {
+
+			// Note the current inspection, otherwise we could end up
+			// in a type-loop and a resulting StackOverflowException
+			if( inspectedTypes.Contains(type)) {
+				return;
+			}
+			inspectedTypes.Add( type );
+
+			if( type.IsTypeMarkedWith( customerStateAttribute ) ) {
+				var location = root.Identifier.GetLocation();
+
+				context.ReportDiagnostic( Diagnostic.Create(
+					Diagnostics.SingletonDependencyHasCustomerState,
+					location
+				) );
+			}
+
+			foreach( ISymbol member in type.GetMembers() ) {

```

Comment:
This will pick up `static` members but I think that's fine. It will also pick up "implicit" members, don't remember what they are but I'll chec tomorrow.

Note that `GetMembers()` will return fewer things if the type is defined in another assembly (e.g. pretty sure it doesn't return the `private` stuff?) This could lead to a situation where a public type XYZ that deep down uses a `[CustomerState]` thing might look safe in a different assembly if XYZ is not annotated.

Perhaps we should also enforce that all `public` things that hold a `[CustomerState]` deep down also be `[CustomerState]`. That would ensure we could do cross-assembly analysis I think.